GPII-1239: Work around reuse of PouchDB between test cases...

the-t-in-rtf opened the pull request:
Because memdown now has a global cache you can only ever have one instance available at a time.  This results in data spilling over between test runs.  This pull request adds tests for this and workaround code that scrubs all database content when the component is destroyed.

See [GPII-1239](https://issues.gpii.net/browse/GPII-1239) for background.

duhrer committed:
GPII-1239:  Added test harness and tests to demonstrate problem with data persisting across multiple test runs.

duhrer committed:
GPII-1239:  Committed work in progress on cleaning the instance at the end of a run and waiting for it to be cleaned in the next run.

duhrer committed:
GPII-1239:  Reverted previous "scrub and wait" approach in favor of simpler workaround.

duhrer committed:
GPII-1239: Updated test data to once again throw an error when confronted with an unclean Pouch database.

duhrer committed:
GPII-1239: Moved `gpii-express` dependency out of dev block, otherwise everyone using the package will have to add `gpii-express` as a dependency on their own.

duhrer committed:
GPII-1239:  Added code to scrub existing databases when the component is destroyed.

duhrer committed:
GPII-1239:  Converted "scrubbing" code to use promises instead of callbacks.

amb26 commented on src/js/pouch.js:
```diff
@@ -25,42 +46,79 @@ gpii.pouch.init = function (that) {
-/*
-    The "databases" option is expected to be an array keyed by dbName, with options to control whether data is loaded or not, as in:
+gpii.pouch.transformRecord = function (record) {
+    // We cannot use "that" or its options here because we have already been destroyed by the time this function is called.
```
This is problematic - since the expectation is that once a component has been destroyed, that all associated resources have been cleaned up. I'm not quite sure how the current test fixtures will be working, since it seems likely that one construction of pouch will interfere with the next one. I will try doing some testing to investigate. Given Infusion's model for component destruction is currently synchronous, I'm not sure there is a very clean way out. We'll have to talk a bit more.

amb26 replied:
So, the following transcript (with a few extra messages) indicates that your current implementation is just working through sheer luck - the extra delay of starting the express server on the second run is probably enough to delay it just long enough that the database has been cleaned:

12:27:57.837:  Successfully queued test Testing initial pouch load...
12:27:57.842:  Successfully queued test Testing second pouch load...
12:27:58.356:  Express server listening on port 6792
12:27:58.359:  Express started...
12:27:58.379:  Sending a GET request to: /sample/ on port 6792
jq:  Test concluded - Test name "Testing initial pouch load...": 1/1 passed - PASS
12:27:58.982:  Destroying pouchDB database sample
12:27:59.008:  Express stopped...
12:27:59.013:  Express server listening on port 6792
12:27:59.016:  Express started...
12:27:59.022:  Beginning to scan database data for destruction
12:27:59.031:  Deleted existing data from database 'sample'...
12:27:59.042:  Sending a GET request to: /sample/ on port 6792
12:27:59.062:  Compacted existing database 'sample'...
12:27:59.076:  Destroying pouchDB database sample
12:27:59.079:  Express stopped...
jq:  Test concluded - Test name "Testing second pouch load...": 1/1 passed - PASS
12:27:59.090:  Beginning to scan database data for destruction
jq:  ***************
jq:  All tests concluded: 2/2 total passed in 2329ms - PASS
jq:  ***************

See comment above for suggestion about making this work properly

amb26 commented on src/js/pouch.js:
```diff
@@ -25,42 +46,79 @@ gpii.pouch.init = function (that) {
     that.expressPouchdb = expressPouchdb(MemPouchDB, { configPath: pouchConfigPath });
 
-    //that.expressPouchdb.couchLogger.setFile(logPath);
+    if (PouchDB.isBeingCleaned) {
```
This is in theory a good idea, but it seems that you never set this flag.
Rather than failing immediately if pouch is busy (which looks like a virtual certainty given the overall workflow) what I suggest is that you store a promise in isBeingCleaned rather than simply a boolean - and if you discover anything there, you simply delay this function's own promise resolution until it has resolved.

the-t-in-rtf replied:
Yes, that was left over from a previous attempt at a similar approach.  I will replace that with the new mechanism as suggested.

the-t-in-rtf replied:
Done (see latest commit).

duhrer committed:
GPII-1239:  Added checks to wait for a previous run to finish its cleanup if needed.

the-t-in-rtf commented:
The updated sequence of events seems reasonable now.

```
13:47:41.775:  Registering module infusion from path /Users/duhrer/Source/rtf/gpii-pouchdb/node_modules/infusion
13:47:41.818:  Registering module kettle from path /Users/duhrer/Source/rtf/gpii-pouchdb/node_modules/kettle
13:47:44.510:  Created new instantiator with id 40kzw8cd-12 in order to operate on component gpii.pouch.tests.reload.environment
13:47:44.519:  Successfully queued test Testing initial pouch load...
13:47:44.520:  Successfully queued test Testing second pouch load...
13:47:46.022:  No previous run detected. Continuing with the normal startup...
13:47:46.551:  Express server listening on port 6792
13:47:46.551:  Express started...
13:47:46.566:  Sending a GET request to: /sample/ on port 6792
jq:  Test concluded - Test name "Testing initial pouch load...": 1/1 passed - PASS
13:47:49.011:  Waiting for the last run to finish its cleanup...
13:47:49.012:  Express stopped...
13:47:49.018:  Express server listening on port 6792
13:47:49.018:  Express started...
13:47:49.021:  Deleted existing data from database 'sample'...
13:47:49.039:  Compacted existing database 'sample'...
13:47:49.041:  Sending a GET request to: /sample/ on port 6792
13:47:49.055:  Express stopped...
13:47:49.057:  Deleted existing data from database 'sample'...
13:47:49.059:  Compacted existing database 'sample'...
```

duhrer committed:
GPII-1239: Added comment about preferred `db.destroy` method, which we apparently cannot use.

amb26 commented on tests/lib/assembleurl.js:
```diff
@@ -0,0 +1,22 @@
+
+
+fluid.registerNamespace("gpii.pouch.tests");
+gpii.pouch.tests.assembleUrl = function (baseUrl, path, that) {
```
linting failure here on "that" unused

amb26 replied:
Linting still remains - also could you please add an "all-tests.js" driver at top level which runs the full test suite - Thanks

the-t-in-rtf replied:
Done, and working on the all-tests driver.

amb26 commented on src/js/pouch.js:
```diff
@@ -25,42 +46,93 @@ gpii.pouch.init = function (that) {
 
-    //that.expressPouchdb.couchLogger.setFile(logPath);
+    var waitForCleaning = PouchDB.isBeingCleaned ? PouchDB.isBeingCleaned : fluid.promise();
+    waitForCleaning.then(function () {
```
Slightly cleaner as well as performing proper cleanup (see below comment) - something like

var initWork = function () {
  delete PouchDB.isBeingCleaned;

....
if (PouchDB.isBeingCleaned) {
  log
    PouchDB.isBeingCleaned.then(initWork);
   }
else {
 log
 initWork();
}

the-t-in-rtf replied:
Done.

amb26 commented on src/js/pouch.js:
```diff
@@ -25,42 +46,93 @@ gpii.pouch.init = function (that) {
+    );
+
+    // Make sure that the next instance of pouch knows to wait for us to finish cleaning up.
+    PouchDB.isBeingCleaned = when.all(promises);
```
To head off similar problems in future, add a hard failure here if PouchDB.isBeingCleaned currently holds a value - note that your current implementation would only be good for 2 tests since you had not performed any cleanup of the cleanup marker :)

the-t-in-rtf replied:
Done.

duhrer committed:
NOJIRA:  Added postnatal call to `grunt deduce-infusion`.

duhrer committed:
GPII-1239:  Tightened up cleanup code and startup checks based on pull request review.

amb26 commented on src/js/pouch.js:
```diff
@@ -1,21 +1,42 @@
+// and populated with the contents of "../tests/data/records.json".
+//
+// NOTE:
+//   This module has a serious and non-obvious limitation, in that only one instance of PouchDB is created.  This
```
Looks like this comment should be updated in the light of the new "cleaning" facility

amb26 commented on package.json:
```diff
@@ -3,25 +3,28 @@
   "private": true,
   "scripts": {
-    "test": "node src/tests/all-tests.js"
+    "test": "node src/tests/all-tests.js",
```
This test driver does not exist!

the-t-in-rtf replied:
I refactored `basic-tests.js` to use a `testEnvironment` and `testCaseHolder`.  Once I did this, it was trivial to just require both tests from an all-tests module.

duhrer committed:
GPII-1239:  Fixed "lint" error (unused variable).

duhrer committed:
GPII-1239:  Refactored "basic" tests so that the entire test harness can be run from a single `all-tests.js` module, and updated `package.json` to use the new file.
